flips/20201310-storage-fees.md

```diff
@@ -0,0 +1,316 @@
+| **Author(s)** | Janez Podhostnik (janez.podhostnik@dapperlabs.com)   |
+| **Sponsor**   | Janez Podhostnik (janez.podhostnik@dapperlabs.com)   |
+| **Updated**   | 2020-10-13                                           |
+
+## Objective
+
+Limit the maximum amount of storage each account has, to prevent storing huge amount of data on chain.
+
+A minimum amount of storage will be provided to each account when it is created. More storage can then be purchased by using the **StorageFeesContract**. Storage can also be refunded using the same contract.
+
+## Motivation
+
+If execution state grows to large execution node will no longer be able to hold it in memory to calculate the next state. When this will happen the execution node requirements (and the nodes them selves) will need to be updated (and will cost more to run).
+
+Limiting the amount of storage an account can have will prevent the execution state growing because of two reasons:
+
+- attacks storing large amounts of data to the blockchain with the explicit intent to slow it down or crash it
+- developers not considering storage costs in their smart contracts (because they don't need to)
+
+## User Benefit
+
+This will have the long term benefit of reducing growth of the execution state size, and thus reducing growth of the execution node requirements. 
+
+Prescribing a reasonable minimum amount of storage will still allow almost everyone to deploy and use smart contracts normally. 
+
+## Design Proposal
+
+### Overview
+
+#### Account Storage Tracking (Execution Environment)
+
+- Every account should have a `storage_capacity` and `storage_used` of type uint64 and unit of byte
+- Minimum amount of `storage_capacity` for each account is 100KB (as defined by the **StorageFeesContract**)
+- `storage_used` is the sum of data stored on ledger (sum of register value byte size owned by this account) (register keys are not part of storage used)
+- on every transaction `storage_used` should be updated
+- both storage_capacity and storage_used is only controlled by the **StorageFeesContract** smart contracts and no user has access to it.
+- If a transaction tries to store more data than the limit, the tx should fail.
+
+#### StorageFeesContract (part of service account)
+
+- is a smart contract that holds the deposits and provides functionality for accounts to acquire or release `storage_capacity`.
+- the smallest unit of storage purchase is initially set to 10KB but can change as a parameter inside the smart contract.
+- price of storage is 1KB for 1mF.
+- accounts can purchase more storage by calling a method on the StorageFeesContract.
+- one account can purchase storage for another account.
+- accounts can release storage (note that each account can't have less than 100KB, at a cost of 0.1F) and get back the money. This feature should be disabled using a boolean initially to prevent attacks and spam transactions to make money by storage deposits.
```

Comment:
Definitely going to add that. 

By the logic described below (execution logic) the last step of the transaction pipeline would check if storage is over capacity, which it would be, so tx would fail.

Reply:
reworded this and should be clearer now